### DISCUSSION TITLE
DAI 2023-2024 - Java IOs benchmarking - Jeremiah Steiner

### DISCUSSION THREAD

**Discussion by BetterBlood** [heig-vd-dai-course/heig-vd-dai-course#87] [Show and tell]
[commit](https://github.com/BetterBlood/heig-vd-dai-course-heig-vd-dai-course-java-ios-practical-content-template/commit/ee3d3a958008c82514e73a589e870f3081161405)
| Type | Opération | Buffer | Temps (ms) | Temps (ns) |
|------|-----------|--------|------------|------------|
| Writing | binary_1B.bin | non-buffered | 1 | 1480300 |
| Reading | binary_1B.bin | non-buffered | 0 | 389600 |
| Writing | binary_1B.bin | buffered | 1 | 1102100 |
| Reading | binary_1B.bin | buffered | 0 | 432300 |
| Writing | text_1B.txt | buffered | 1  | 1923200 |
| Reading | text_1B.txt | buffered | 0  | 660100 |
| Writing | text_1B.txt | non-buffered | 1   | 1279600 |
| Reading | text_1B.txt | non-buffered | 0  | 600600 |
| Writing | binary_1KiB.bin | non-buffered | 21  | 21197800 |
| Reading | binary_1KiB.bin | non-buffered | 9   | 9765700 |
| Writing | binary_1KiB.bin | buffered | 1  | 1158300 |
| Reading | binary_1KiB.bin | buffered | 1  | 1487900 |
| Writing | text_1KiB.txt | buffered | 2  | 2308200 |
| Reading | text_1KiB.txt | buffered | 1  | 1553600 |
| Writing | text_1KiB.txt | non-buffered | 8 | 8561700 |
| Reading | text_1KiB.txt | non-buffered | 4  | 4203900 |
| Writing | binary_1MiB.bin | non-buffered | 6918  | 6918317600 |
| Reading | binary_1MiB.bin | non-buffered | 6076  | 6076323000 |
| Writing | binary_1MiB.bin | buffered | 32| 32630400 |
| Reading | binary_1MiB.bin | buffered | 47 | 47842000 |
| Writing | text_1MiB.txt | buffered | 78 | 78942500 |
| Reading | text_1MiB.txt | buffered | 110  | 110107800 |
| Writing | text_1MiB.txt | non-buffered | 203  | 203484500 |
| Reading | text_1MiB.txt | non-buffered | 173 | 173512000 |
| Writing | binary_5MiB.bin | non-buffered | 33086  | 33086780500 |
| Reading | binary_5MiB.bin | non-buffered | 30096  | 30096965000 |
| Writing | binary_5MiB.bin | buffered | 226 | 226607700 |
| Reading | binary_5MiB.bin | buffered | 226 | 226585800 |
| Writing | text_5MiB.txt | buffered | 334  | 334934100 |
| Reading | text_5MiB.txt | buffered | 382 | 382172900 |
| Writing | text_5MiB.txt | non-buffered | 577  | 577779600 |
| Reading | text_5MiB.txt | non-buffered | 440 | 440180400 |

Which type of stream is the most efficient for each use case? 

|size|type|
|------|-----------|
|1 B| not significant|
|1 KiB | buffered binary|
|1 MiB | buffered binary|
|5 MiB | buffered binary|

Why is it more efficient than the other types of streams? cause it optimise the information transfer using buffers
What is the difference between binary data and text data? a char could take more than a byte
What is a character encoding? a meta information, let us know how to read bytes to understand them the right way
Why is this methodology important? to prevent loosing time on execution, to optimise our application

ChatGPT used for creating the tables using this prompt: 
Récupère les données dans le texte ci-dessous, et formate certaines informations en tableau en format markdown. Les informations nécessaire sont : si le fichier est en binaire ou en texte, si l'utilisation d'un buffer est effectuée ou non, si c'est une lecture ou une écriture, le nombre de ms et le nombre de ns. Dans les données ci-dessous les infos sont groupés par deux lignes, je veux donc dans le tableau uns ligne pour chaque deux ligne du texte ci-dessous. Donne directement le .md
"Writing 1 bytes to binary_1B.bin with a non-buffered stream...
Elapsed time: 1ms / 1480300ns 
{...}
Reading text_5MiB.txt with a non-buffered stream...
Elapsed time: 440ms / 440180400ns"

**Comment by ludelafo** [comment]
Great, thank you!

I guess the repository is private as I cannot access it. If you add us to your project or make it public, we can give you a feedback.

Good usage of ChatGPT.

**Reply by BetterBlood** [reply]
That's right, forget about this point, I've just added you two to my repo, I would appreciate feedback.

**Reply by ludelafo** [reply]
Thank you!

Everything seems fine by me. One thing you could improve is the usage of the `try-with-resources` type of `try() { ... }` as mentioned in the course or in the following tutorial: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

The other thing you could improve is to add the `.bin` and `.txt` files to your gitignore files as these are generated by your application and take some space on your repository.

Other than this, everything looks good to me!

**Reply by BetterBlood** [reply]
That's done, thanks for the feedback.

**Reply by ludelafo** [reply]
You're welcome!
